File: CS/Inline_Unobtrusive/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using Inline_Unobtrusive.Models;

namespace Inline_Unobtrusive.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View(GetData());
        }

        public ActionResult GridViewPartial() {
            return PartialView(GetData());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult InlineEditingAddNewPartial(Customer customer) {
            if (ModelState.IsValid) {
                try {
                    //insert data in the data source
                } catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return PartialView("GridViewPartial", GetData());
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult InlineEditingUpdatePartial(Customer customer) {
            if (ModelState.IsValid) {
                try {
                    //update data in the data source
                } catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return PartialView("GridViewPartial", GetData());
        }

        public static IEnumerable<Customer> GetData() {
            return Enumerable.Range(0, 10).Select(i => new Customer() { ID = i, City = "City " + i, Name = "Customer " + i });
        }
    }

}

File: VB/Inline_Unobtrusive/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web.Mvc
Imports Inline_Unobtrusive.Models

Namespace Inline_Unobtrusive.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View(GetData())
		End Function

		Public Function GridViewPartial() As ActionResult
			Return PartialView(GetData())
		End Function

		<HttpPost, ValidateInput(False)> _
		Public Function InlineEditingAddNewPartial(ByVal customer As Customer) As ActionResult
			If ModelState.IsValid Then
				Try
					'insert data in the data source
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If
			Return PartialView("GridViewPartial", GetData())
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function InlineEditingUpdatePartial(ByVal customer As Customer) As ActionResult
			If ModelState.IsValid Then
				Try
					'update data in the data source
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If
			Return PartialView("GridViewPartial", GetData())
		End Function

		Public Shared Function GetData() As IEnumerable(Of Customer)
			Return Enumerable.Range(0, 10).Select(Function(i) New Customer() With {.ID = i, .City = "City " & i, .Name = "Customer " & i})
		End Function
	End Class

End Namespace